crypto/threads_pthread.c

```diff
@@ -188,7 +195,7 @@ int CRYPTO_THREAD_compare_id(CRYPTO_THREAD_ID a, CRYPTO_THREAD_ID b)
 
 int CRYPTO_atomic_add(int *val, int amount, int *ret, CRYPTO_RWLOCK *lock)
 {
-# if defined(__GNUC__) && defined(__ATOMIC_ACQ_REL)
+# if defined(__GNUC__) && defined(__ATOMIC_ACQ_REL) && !defined(BROKEN_CLANG_ATOMICS)
     if (__atomic_is_lock_free(sizeof(*val), val)) {
         *ret = __atomic_add_fetch(val, amount, __ATOMIC_ACQ_REL);
```

Comment:
Is it possible to make atomic operations work on these OS/X versions?  I.e. wrapping this:

```c
#  ifdef BROKEN_CLANG_ATOMICS
        *ret = /* OS/X version */
#  else
        *ret = __atomic_or_fetch(val, op, __ATOMIC_ACQ_REL);
#  endif
```

This will prevent the code falling back to the locking alternative.

Reply:
I assume similar fix is needed for Android: https://github.com/openssl/openssl/issues/18059

Reply:
That will be a different fix, it looks like an extra library is required not the wrong number of arguments to the intrinsic.

Reply:
@paulidale the issue is much deeper and I don't think that it is possible. I'll explain it in comment to keep everything in one place.